Mounting dependencies on paths specified by keys

teetone opened the pull request:
Resolves #1115 

- Support `.`, which will mount all content on the top-level of the dependencies path
- Support custom paths via keys
- We currently don't support paths to be an ancestor of another path since we mount file-systems to be read-only

teetone committed:
mounting

percyliang reviewed (commented):
Logic looks good overall!  I'd try to make the code a bit more modular - the logic for attaching mounting a path within a dependency to some path versus how we interpret `.`.  Note that if we wanted to support having keys `foo` and `foo/bar`, then ideally we'd use the same logic for how we deal with `.`: just mount everything under `foo`.

teetone committed:
initial cleanup

teetone committed:
some refactoring and fixed the case for shared FS

teetone committed:
added new lines

teetone commented:
> Logic looks good overall! I'd try to make the code a bit more modular - the logic for attaching mounting a path within a dependency to some path versus how we interpret `.`. Note that if we wanted to support having keys `foo` and `foo/bar`, then ideally we'd use the same logic for how we deal with `.`: just mount everything under `foo`.

@percyliang I fixed the issue I was having with shared filesystems and added `test2` to  the shared filesystems run of our Travis build. I also did some refactoring and made the logic a bit more modular.

teetone committed:
resolve merge conflicts

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1115-mount

teetone committed:
keep track of paths to remove while mounting

yipenghe committed:
Merge branch 'master' into 1115-mount

candicegjing committed:
Merge branch 'master' into 1115-mount

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1115-mount

teetone committed:
get rid of invalid test since keys can be paths

teetone committed:
Merge branch '1115-mount' of https://github.com/codalab/codalab-worksheets into 1115-mount

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1115-mount

teetone committed:
support absolute path + add test

teetone committed:
resolve merge conflicts

teetone commented:
@percyliang @candicegjing I modified the PR to support absolute paths. When a key is an absolute path, I ended up removing the leaf to clean up the dependency that is mounted.

percyliang reviewed (approved):
Great!

teetone committed:
Merge branch 'master' of https://github.com/codalab/codalab-worksheets into 1115-mount

teetone committed:
cleanup

teetone committed:
Merge branch 'master' into 1115-mount
